pyccel/ast/basic.py

```diff
@@ -16,6 +19,11 @@ class Basic(sp_Basic):
     """Basic class for Pyccel AST."""
     _fst = None
 
+    def __new__(cls, *args, **kwargs):
+        hashable_args  = [a if not isinstance(a, list) else tuple(a) for a in args]
+        hashable_args += [a if not isinstance(a, list) else tuple(a) for a in kwargs.values()]
+        return sp_Basic.__new__(cls, *hashable_args)
```

Comment:
I am not convinced that creating a list of input arguments (ordered) from a dictionary of keyword arguments (not necessarily ordered) is useful. Further, we could get unexpected behavior.

Reply:
ordering a dictionary should not be a problem, because the order doesn't matter from the beginning

Reply:
Dictionaries are ordered for python 3.7+, I am not using the `kwargs` passed to `sp_Basic` (unless an class with args only is called using keywords I suppose) so I don't think we could get any unexpected behaviour.

I did this as sympy has internal singleton behaviour. I have not looked into it in detail, but it seems that if the `_args` attribute is the same for 2 instances of a class then they are considered to be the same object. Thus when we modify one, the other changes too which may not be what we wanted.

If anyone has a better solution for this then I would be happy to implement it (@saidctb ?).

Otherwise I can try to just call `sp_Basic.__new__(cls)` and see if anything breaks. It may be that the singleton behaviour only works for certain sympy classes

Reply:
> Otherwise I can try to just call sp_Basic.__new__(cls) and see if anything breaks

This does not work for any classes which use `_args` instead of mutable objects

Reply:
It seems that removing line 24 does not break anything. I will remove this line

Reply:
Let me help understand this. We are passing a list of hash-able arguments to `sympy.Basic.__new__` in the hope that Sympy's caching mechanism will not treat them as the same object. But if the arguments are the same we are still in trouble, right?

Reply:
Yes I think so. But this has always been the case so we are not generating bugs here that we did not have before

Reply:
That is why I wanted to also store the kwargs. The more information is stored, the less chance there is of the `_args` matching

Reply:
OK, let's keep things as they are after the last commit. In any case this confusion will be resolved as soon as we remove all dependencies on Sympy.